src/Web/XHR/Types.purs

```diff
@@ -0,0 +1,94 @@
+module Web.XHR.Types
+  ( XHR
+  , XMLHttpRequest
+  , XMLHttpRequestUpload
+  , ReadyState(..)
+  , ResponseType
+  , FormData
+  , string
+  , blob
+  , document
+  , unsafeResponseType
+  , xmlHttpRequestToEventTarget
+  , xmlHttpRequestUploadToEventTarget
+  ) where
+
+import Prelude
+
+import Control.Monad.Eff (kind Effect)
+import DOM.Event.Types (EventTarget)
+import DOM.File.Types (Blob)
+import DOM.Node.Types (Document)
+import Data.Enum (class BoundedEnum, class Enum, Cardinality(..), defaultPred, defaultSucc, fromEnum, toEnum)
+import Data.Maybe (Maybe(..))
+import Type.Proxy (Proxy)
+import Unsafe.Coerce (unsafeCoerce)
+
+foreign import data XHR :: Effect
+
+-- | `XMLHttpRequest`s are indexed by their `ResponseType`
+foreign import data XMLHttpRequest :: Type -> Type
+
+foreign import data XMLHttpRequestUpload :: Type
+
+foreign import data FormData :: Type
+
+data ReadyState
+  = ReadyStateUnsent
+  | ReadyStateOpened
+  | ReadyStateHeadersReceived
+  | ReadyStateLoading
+  | ReadyStateDone
+
+derive instance eqReadyState :: Eq ReadyState
+derive instance ordReadyState :: Ord ReadyState
+
+instance boundedReadyState :: Bounded ReadyState where
+  bottom = ReadyStateUnsent
+  top = ReadyStateDone
+
+instance enumReadyState :: Enum ReadyState where
+  succ = defaultSucc toEnum fromEnum
+  pred = defaultPred toEnum fromEnum
+
+instance boundedEnumReadyState :: BoundedEnum ReadyState where
+  cardinality = Cardinality 5
+  toEnum = toEnumReadyState
+  fromEnum = fromEnumReadyState
+
+newtype ResponseType res = ResponseType String
+
+string :: ResponseType String
+string = ResponseType ""
+
+blob :: ResponseType Blob
+blob = ResponseType "blob"
+
+document :: ResponseType Document
+document = ResponseType "document"
+
+unsafeResponseType :: forall res. Proxy res -> String -> ResponseType res
+unsafeResponseType _ = ResponseType
```

Comment:
One last thing... is having / exporting this a good idea?

Reply:
The reason I didn't is because there are actually more response types (JSON, Array Buffer), but we don't have canonical types for them.

Reply:
JSON doesn't work pretty much, so excluding that one is fair enough I think. ArrayBuffer we do have `purescript-arraybuffer-types`.

The fact that `res` is later `unsafeCoerce`d using this phantom type is what worries me a bit, since it means the problem is not local to the usage of `unsafeResponseType`. Hmmm.

Reply:
I'm fine with closing it off and importing arraybuffer types.